11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+			};
+			const auto srcImageViewCPU = ICPUImageView::create(std::move(viewParams[0]));
+			const auto kerImageViewCPU = ICPUImageView::create(std::move(viewParams[1]));
+
+			// Using asset converter
+			smart_refctd_ptr<nbl::video::CAssetConverter> converter = nbl::video::CAssetConverter::create({ .device = m_device.get(),.optimizer = {} });
+			// We don't want to generate mip-maps for these images (YET), to ensure that we must override the default callbacks.
+			struct SInputs final : CAssetConverter::SInputs
+			{
+				inline uint8_t getMipLevelCount(const size_t groupCopyID, const ICPUImage* image, const CAssetConverter::patch_t<asset::ICPUImage>& patch) const override
+				{
+					return image->getCreationParameters().mipLevels;
+				}
+				inline uint16_t needToRecomputeMips(const size_t groupCopyID, const ICPUImage* image, const CAssetConverter::patch_t<asset::ICPUImage>& patch) const override
+				{
+					return 0b0u;
+				}
+			} inputs = {};
+			inputs.logger = m_logger.get();
+			nbl::asset::ICPUImageView* CPUImageViews[2] = { srcImageViewCPU.get(), kerImageViewCPU.get() };
+			
+
+			// Need to provide patches to make sure we specify SAMPLED to get READ_ONLY_OPTIMAL layout after upload
+			CAssetConverter::patch_t<ICPUImageView> patches[2] =
+			{
+				{
+					CPUImageViews[0],
+					IImage::E_USAGE_FLAGS::EUF_SAMPLED_BIT
+				},
+				{
+					CPUImageViews[1],
+					IImage::E_USAGE_FLAGS::EUF_SAMPLED_BIT
+				}
+			};
+
+			std::get<CAssetConverter::SInputs::asset_span_t<ICPUImageView>>(inputs.assets) = { CPUImageViews, 2 };
+			std::get<CAssetConverter::SInputs::patch_span_t<ICPUImageView>>(inputs.patches) = { patches, 2 };
+			auto reservation = converter->reserve(inputs);
+			const auto GPUImages = reservation.getGPUObjects<ICPUImageView>();
+
+			m_srcImageView = GPUImages[0].value;
+			m_kerImageView = GPUImages[1].value;
+
+			// Give them debug names
+			m_srcImageView->setObjectDebugName("Source image view");
+			m_srcImageView->getCreationParameters().image->setObjectDebugName("Source Image");
+			m_kerImageView->setObjectDebugName("Bloom kernel image view");
+			m_kerImageView->getCreationParameters().image->setObjectDebugName("Bloom kernel Image");
+
+			// The down-cast should not fail!
+			assert(m_srcImageView);
+			assert(m_kerImageView);
+
+			// Required size for uploads
+			auto srcImageDims = m_srcImageView->getCreationParameters().image->getCreationParameters().extent;
+			auto kerImageDims = m_kerImageView->getCreationParameters().image->getCreationParameters().extent;
+			// Add a bit extra because EXR has alpha
+			uint32_t srcImageSize = srcImageDims.height * srcImageDims.width * srcImageDims.depth * (CHANNELS + 1) * sizeof(float32_t);
+			uint32_t kerImageSize = kerImageDims.height * kerImageDims.width * kerImageDims.depth * (CHANNELS + 1) * sizeof(float32_t);
+
+			m_utils = make_smart_refctd_ptr<IUtilities>(smart_refctd_ptr(m_device), smart_refctd_ptr(m_logger), srcImageSize, srcImageSize + kerImageSize);
```

Comment:
stop it with the arithmetic, you're giving people the impression we can't do chunked uploads, use default upload and download streaming  buffer sizes